### PR TITLE
SAK-43768 SAMIGO: Randomize question pools with option: A student's questions are randomized once for all submissions

### PR DESCRIPTION
**Pull request by PabloArrz**
https://sakaiproject.atlassian.net/browse/SAK-43768

### CODE REVIEW

**Review by mpellicer — CHANGES_REQUESTED** [review]
This PR could introduce backward compatibility issues, @vacercon may expand on that.

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/delivery/DeliveryActionListener.java [hunk 1]
```diff
@@ -2863,6 +2863,8 @@ private String getTimerId(DeliveryBean delivery){
   
   private long getSeed(SectionDataIfc sectionData, DeliveryBean delivery, long userSeed) {
 	  long seed = userSeed;
+	  String sectionRandomizationSeed = sectionData.getSectionMetaDataByLabel(SectionDataIfc.RANDOMIZATION_SEED);
+	  if (StringUtils.isNotBlank(sectionRandomizationSeed)) seed += Long.parseLong(sectionRandomizationSeed);
```
Comment (mpellicer) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/delivery/DeliveryActionListener.java:2867]: Please, include a comment.

// If the section has a previous seed we must use it to use the same order.

Also use brackets here:

if (StringUtils.isNotBlank(sectionRandomizationSeed)) {
   seed += Long.parseLong(sectionRandomizationSeed);
}

File: samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/facade/PublishedAssessmentFacadeQueries.java [hunk 2]
```diff
@@ -346,6 +347,7 @@ public Set preparePublishedSectionMetaDataSet(
 					sectionMetaData.getEntry());
 			h.add(publishedSectionMetaData);
 		}
+		h.add(new PublishedSectionMetaData(publishedSection, SectionDataIfc.RANDOMIZATION_SEED, String.valueOf(UUID.randomUUID().hashCode())));
```
Comment (mpellicer) [inline samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/facade/PublishedAssessmentFacadeQueries.java:350]: Please include this comment:

// Persist the random seed in the section to use it and preserve the order.

**Review by mpellicer — CHANGES_REQUESTED** [review]
Just improve the documentation and include brackets.

**Comment by mpellicer** [comment]
Well done @PabloArrz ! Clean and neat.

**Comment by PabloArrz** [comment]
@mpellicer I have improved the documentation and include the brackets.

**Comment by mpellicer** [comment]
@josecebe WDYT? Ready to go?

**Review by mpellicer — APPROVED** [review]

**Comment by josecebe** [comment]
> @josecebe WDYT? Ready to go?

It works nice, but we will test it deeply after it gets merged, I saw there are more seed references in the code, but maybe is not necessary to modify those seeds to keep it working properly.

**Comment by mpellicer** [comment]
Thanks for the feedback

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/delivery/DeliveryActionListener.java [hunk 3]
```diff
@@ -2863,6 +2863,11 @@ private String getTimerId(DeliveryBean delivery){
   
   private long getSeed(SectionDataIfc sectionData, DeliveryBean delivery, long userSeed) {
 	  long seed = userSeed;
+	  // If the section has a previous seed we must use it to use the same order.
+	  String sectionRandomizationSeed = sectionData.getSectionMetaDataByLabel(SectionDataIfc.RANDOMIZATION_SEED);
+	  if (StringUtils.isNotBlank(sectionRandomizationSeed)) {
+	      seed += Long.parseLong(sectionRandomizationSeed);
+	  }
```
[the remark this answers is not in the document]
Comment (mpellicer) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/delivery/DeliveryActionListener.java:2870]: @josecebe Could you clarify this? Thanks

[the remark this answers is not in the document]
Comment (bbailla2) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/delivery/DeliveryActionListener.java:2870]: Here's my understanding of the intention:
When a new Section is created, then we insert a seed into the section metadata. This particular snippet checks for the existence of a randomization seed in the metadata, and if the result is non-blank, then it's added to the seed - this bit looks fine to me

**Pull request merged**